### ISSUE TITLE
Java API - race condition with state listeners

### ISSUE THREAD

**Issue by cverges** [dnp3/opendnp3#27]
The process to define a stack is:

```
Channel channel = new DNP3Manager().addTCPClient(...);
channel.addStateListener(listener);

Master master = channel.addMaster(...);
master.addStateListener(listener);
```

Since execution starts immediately after creation, there is an uncontrollable time delay between creating the object and registering the state listeners at both the Channel and Master level.

**Comment by jadamcrain** [comment]
The registration of a state listener should cause the current state to be sent to the listener immediately. Is this not happening?

**Comment by cverges** [comment]
It is happening as you say, but there is no way to control the amount of time between creation of the stack and registration of the listener.  So if the scheduler gets busy, it is theoretically possible for the channel/stack to undergo more than one state transition such that all but the last would be lost to the listener.

**Comment by jadamcrain** [comment]
Ok. I will think about this more carefully when I'm not tired ;).

Thanks for opening the ticket.

Has this behavior been observed for just theorized?

**Comment by cverges** [comment]
Just theorized at this point.  Haven't started loading up my server yet, but I could easily foresee it under heavy loads  (low memory, high cpu, few threads, etc.)

**Comment by jadamcrain** [comment]
It may be best just to limit these state callback definitions to construction. Thoughts?

**Comment by cverges** [comment]
This makes sense, but only if we have a master.start() that can be triggered independently from construction.  Construction shouldn't be seen as an "auto-run" step, basically.

**Comment by jadamcrain** [comment]
Looked into this.  Yes, you can miss events early on, but you can never lose current state.  The registrations happens on the executor assigned to the channel and it always queues the current state callback on registration.

Either way, it's fixed in 2.0.x because you can explicitly control when the stack starts.
